Add support for filecache in utility functions

jnspitale opened the pull request:
Several utility functions have been modified to accept a `FCPath` argument.  

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Public utilities now accept multiple path formats (strings, Path objects, and custom file-path objects), improving support for local and remote/cloud paths.
  * File metadata (size, hash, record counts) and reads operate consistently across path types; modification times are returned in standardized ISO/Zulu form.

* **Documentation**
  * Docstrings updated to reflect expanded input types and behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jnspitale committed:
Add support for filecache in utility functions

jnspitale committed:
added warning about FILE_TIME and FILE_ZULU mod times

jnspitale committed:
FILE_PATH and FILE_ZULU warning contingent on file being remote

jnspitale committed:
Clean up

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -1055,6 +1056,7 @@ def FILE_BYTES(filepath):
             int: The size in bytes of the file.
         """
 
+        filepath = FCPath(filepath).retrieve()
```
It's generally a bad idea to reuse a variable name for two purposes. The `filepath` passed in is a URL, while the `filepath` from the `retrieve` is a local path. It would be better to call it `local_path`. This applies to all other applications below.

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -857,7 +857,8 @@ def BASENAME(filepath):
 
-        return os.path.basename(filepath)
+        return FCPath(filepath).name
+#        return os.path.basename(filepath)
```
Remove dead comment

rfrenchseti commented on pdstemplate/__init__.py:
The type of the parameter needs to be updated to say `Path | FCPath | str` everywhere this change is made.

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -1125,6 +1129,15 @@ def FILE_TIME(filepath):
+        ##################################################################################
+        local_filepath = FCPath(filepath).retrieve()
+        # Only warn for remote files where cached time is returned
+        if local_filepath != filepath:
```
I need to add a small new method to `FCPath` so you can access the path's FileCache. Then you would say:

```
fc_path = FCPath(filepath)
local_filepath = fc_path.retrieve()
if not fc_path.filecache.is_time_sensitive and not fc_path.is_local():
    (warn)
```

This will also fix the code rabbit complaint.

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -1140,6 +1153,15 @@ def FILE_ZULU(filepath):
+        ##################################################################################
+        local_filepath = FCPath(filepath).retrieve()
+        # Only warn for remote files where cached time is returned
+        if local_filepath != filepath:
```
Ditto above

rfrenchseti reviewed (changes requested):
Please address the Code Rabbit comments as well.

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -1140,6 +1153,15 @@ def FILE_ZULU(filepath):
+            logger = get_logger()
+            logger.warning(f'FILE_TIME reflects the cached time: {filepath}', force=True)
+        ##################################################################################
+#        timestamp = FCPath(filepath).stat().st_mtime  ### not implemented in FCPath
```
As discussed elsewhere, this should be `FCPath(filepath).modification_time()`, which will work today for both local and cloud storage locations.

jnspitale replied:
Thanks, the temporary code is replaced.  fc_path.filecache access is no longer needed in this context.

jnspitale committed:
- addressed pdstemplate coderabbit comments
- replaced temporary code in FILE_TIME and FILE_ZULU with call to FCPATH.modification_time()

rfrenchseti commented:
@jnspitale What is the status of this PR? Have you fixed all of the CodeRabbit and my comments and is it ready to re-review? If so, please click the "re-review" icon at the top next to my name.

rfrenchseti commented on pdstemplate/__init__.py:
```diff
@@ -1049,28 +1049,30 @@ def FILE_BYTES(filepath):
+        local_path = FCPath(filepath).retrieve()
         blocksize = 65536
-        with open(filepath, 'rb') as f:
+        with open(local_path, 'rb') as f:
```
This is fine, but to be a little cleaner you can do:

```
fc_path = FCPath(filepath)
with fc_path.open('rb') as f:
```

here and in the other instances below.

jnspitale replied:
why not local_path.open('rb')?

rfrenchseti replied:
There are three ways to do this:

```
#1:
local_path = ...
with open(local_path, 'rb') as f:

#2:
local_path = ...
with local_path.open('rb') as f:

#3:
with FCPath(filepath).open('rb') as f:
```

I think the last one is nicest because it doesn't require an explicit `retrieve`. That hides the internals of filecache. But they are all equivalent.

jnspitale replied:
Ok, I also prefer #3.  Thereis only one retrieve() left in the utility functions, in FILE_BYTES().  Is there a way to get that without downloading the file?

rfrenchseti replied:
Not right now. It's probably not worth putting it in for this one use case. And any time someone needs the number of bytes they also likely need the MD5 checksum, so it will have to be retrieved eventually anyway.

rfrenchseti reviewed (changes requested):
What is the file `test_files/lbl`? It appears to be empty.

jnspitale committed:
Eliminate FCPath retrieves in utility functions

jnspitale committed:
coderabbit
